src/natcap/invest/ndr/ndr_core.pyx

```diff
@@ -183,36 +164,33 @@ def ndr_eff_calculation(
                     global_col, global_row, effective_retention_nodata)
             else:
                 working_retention_eff = 0.0
-                outflow_weight_sum = 0
-                for i in range(8):
-                    outflow_weight = (flow_dir >> (i*4)) & 0xF
-                    if outflow_weight == 0:
-                        continue
-                    outflow_weight_sum += outflow_weight
-                    ds_col = col_offsets[i] + global_col
-                    if ds_col < 0 or ds_col >= n_cols:
+                has_outflow = False
+                for neighbor in mfd_flow_direction_raster.yield_downslope_neighbors(
+                        global_col, global_row, skip_oob=False):
+                    has_outflow = True
```

Comment:
We know this to be `True` right away because `yield_downslope_neighbors` will return empty and not enter here otherwise?

Reply:
Correct, if a pixel has no downslope neighbors (i.e. has a MFD value of 0), then an error is raised below